libraries/rush-lib/src/logic/operations/CacheableOperationPlugin.ts

```diff
@@ -0,0 +1,1107 @@
+                   *   1. Y must be built by the same VM that build X;
+                   *   2. OR, Y must be rebuilt on each VM that needs it.
+                   * Approach 1 is probably the better choice.
+                   */
+                  for (const consumer of record.consumers) {
+                    disjointSet?.union(record, consumer);
+                  }
+                }
+              }
+            },
+            {
+              concurrency: 10
+            }
+          );
+
+          for (const set of disjointSet.getAllSets()) {
+            if (cobuildConfiguration?.cobuildEnabled && cobuildConfiguration.cobuildContextId) {
+              // Get a deterministic ordered array of operations, which is important to get a deterministic cluster id.
+              const groupedRecords: OperationExecutionRecord[] = Array.from(set);
+              Sort.sortBy(groupedRecords, (record: OperationExecutionRecord) => {
+                return record.runner.name;
+              });
+
+              // Generates cluster id, cluster id comes from the project folder and phase name of all operations in the same cluster.
+              const hash: crypto.Hash = crypto.createHash('sha1');
+              for (const record of groupedRecords) {
+                const { associatedPhase: phase, associatedProject: project } = record;
+                if (project && phase) {
+                  hash.update(project.projectRelativeFolder);
+                  hash.update(RushConstants.hashDelimiter);
+                  hash.update(phase.name);
+                  hash.update(RushConstants.hashDelimiter);
+                }
+              }
+              const cobuildClusterId: string = hash.digest('hex');
+
+              // Assign same cluster id to all operations in the same cluster.
+              for (const record of groupedRecords) {
+                const buildCacheContext: IOperationBuildCacheContext =
+                  this._getBuildCacheContextByOperationExecutionRecordOrThrow(record);
+                buildCacheContext.cobuildClusterId = cobuildClusterId;
+              }
+            }
+          }
+        }
+      }
+    );
+
+    hooks.beforeExecuteOperation.tapPromise(
+      PLUGIN_NAME,
+      async (runnerContext: IOperationRunnerContext): Promise<OperationStatus | undefined> => {
+        const { _createContext: createContext } = this;
+        if (!createContext) {
+          return;
+        }
+        const {
+          projectChangeAnalyzer,
+          buildCacheConfiguration,
+          cobuildConfiguration,
+          phaseSelection: selectedPhases
+        } = createContext;
+
+        const record: OperationExecutionRecord = runnerContext as OperationExecutionRecord;
+        const {
+          associatedProject: project,
+          associatedPhase: phase,
+          _operationMetadataManager: operationMetadataManager
+        } = record;
+
+        if (!project || !phase) {
```

Comment:
Other than these two properties, we should add an explicit opt-out of caching behavior, e.g. for operations that are known to have side-effects that are not cacheable.

Reply:
I recommend just adding a `boolean` property `cacheable` to `Operation` that initializes to false, and that the `ShellOperationRunnerPlugin` sets to `true` when it attaches `ShellOperationRunner` instances to operations.

Then we check that property here as well as any others we need.